tools/sof-perf-analyzer.py

```diff
@@ -0,0 +1,253 @@
+    test case due to mtrace is configured in deferred mode. This function consumes
+    those invalid traces from the generator, and return the first valid trace item.
+    '''
+    try:
+        while item := next(trace_item_gen):
+            if int(item.timestamp) == 0:
+                return item
+    except StopIteration:
+        return None
+    return None
+
+def make_trace_item(fileio: TextIO) -> TraceItemGenerator:
+    '''Convert a line of trace in string form into TraceItem object, for example:
+    '[0.349880] <inf> pipe: pipeline_connect: comp:0 0x40000 connect buffer 0 as sink' ->
+    TraceItem(timestamp=0.349880, level='inf', context='pipe', func='pipeline_connect',
+              msg='comp:0 0x40000 connect buffer 0 as sink')
+    '''
+    for line in fileio:
+        # Filter extra lines that are not formal traces, for example, the banner,
+        # which don't contain timestamp and trace level.
+        if match_obj := re.search(r'\[.+\] <(dbg|inf|wrn|err)>', line):
+            # Sometimes, A trace output may be incomplete and mixed with next
+            # trace output in a single line, for example: '[    0.071590] <inf>
+            # pipe[    0.071751] <inf> host_comp: host_get_copy_bytes_normal: comp:1
+            # 0x40003 no bytes to copy, available samples: 0, free_samples: 384'.
+            # Under this circumstance, we ignore the first trace and extract the
+            # second trace from the line. This is done by regarding the end position
+            # of the matched substring as sentinel, timestamp and trace level are
+            # both some specific offset from the sentinel.
+            span_end_pos = match_obj.span()[1]
+            trace_lvl = line[span_end_pos - 4: span_end_pos - 1]
+            timestamp = float(line[span_end_pos - 19: span_end_pos - 7].strip())
+
+            # The rest after removing timestamp and log level
+            rest = line[span_end_pos + 1:].split(': ')
+            ctx, func = rest[0:2]
+            msg = ': '.join(rest[2:]).strip()
+            yield TraceItem(timestamp, trace_lvl, ctx, func, msg)
+
+def process_trace_file():
+    '''The top-level caller for processing the trace file'''
+    with open(args.filename, 'r', encoding='utf8') as file:
+        trace_item_gen = make_trace_item(file)
+        trace_prev = get_first_valid_trace(trace_item_gen)
+        if trace_prev is None:
+            raise Exception('No valid trace in provided file')
+        for trace_curr in trace_item_gen:
+            # pylint: disable=W0603
+            global ts_shift
+            old_ts_shift = ts_shift
+            # On wrap happened, the timestamp of current trace should be much more smaller
+            # than the previous one. In practice, it is possible that the timestamp of
+            # current trace is slightly smaller than the previous one, this could be a
+            # bug in SOF. Add a 50s shift to make sure timestamp correction work properly.
```

Comment:
How can we trust performance information provided by the firmware if timestamps are wrong?

Reply:
The timestamp is not wrong, it just wrapped if a test is running too long (> 111.85 seconds). That why we do timestamp correction, add 111.85 seconds at every wrap.